### PR TITLE
JAMES-3534 Identity/set destroy

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 1]
```diff
@@ -663,4 +670,304 @@ trait IdentitySetContract {
...
+           |			{
+           |				"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+           |				"newState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
+           |				"notDestroyed": {
+           |					"$serverIdentitiesId1": {
+           |						"type": "forbidden",
+           |						"description": "User do not have permission to delete IdentityId($serverIdentitiesId1)",
+           |						"properties": [
+           |							"id"
+           |						]
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:779]: I don't think properties field applies for `forbidden` method-level error level.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetDeletePerformer.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        .onErrorRecover(e => IdentityDeletionFailure(unparsedId, e)),
+        maxConcurrency = 5)
+      .collectSeq()
+      .map(IdentityDeletionResults)
+
+  private def delete(unparsedId: UnparsedIdentityId, mailboxSession: MailboxSession): SMono[IdentityDeletionResult] =
+    unparsedId.validate
+      .fold(e => SMono.error(e),
+        id => SMono.fromPublisher(identityRepository.delete(mailboxSession.getUser, Seq(id)))
+          .subscribeOn(Schedulers.elastic())
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetDeletePerformer.scala:73]: Why do we need an elastic scheduler here? This detail  should be handled within `identityRepository.delete`

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala [hunk 3]
```diff
@@ -128,7 +131,19 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
 
   def update(user: Username, identityId: IdentityId, identityUpdate: IdentityUpdate): Publisher[Unit] = customIdentityDao.update(user, identityId, identityUpdate)
 
-  def delete(username: Username, ids: Seq[IdentityId]): Publisher[Unit] = customIdentityDao.delete(username, ids)
+  def delete(username: Username, ids: Seq[IdentityId]): Publisher[Unit] = {
```
Comment (Arsnael) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala:134]: Can remove brackets `{}` here?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 4]
```diff
@@ -663,4 +670,295 @@ trait IdentitySetContract {
...
+           |						"type": "forbidden",
+           |						"description": "User do not have permission to delete IdentityId($defaultServerSetIdentity)"
+           |					}
+           |				}
+           |			},
+           |			"c1"
+           |		]
+           |	]
+           |}""".stripMargin)
+  }
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:963]: Should we add a test to make sure we can't fetch an identity after it has been deleted?

**Comment by quantranhong1999** [comment]
Look like I found a bug at Identity/get.
When getting a random id, that id should be responded in notFound but did not.
Test:
```
   val id = UUID.randomUUID().toString
    val request =
      s"""{
         |  "using": ["urn:ietf:params:jmap:core", "urn:ietf:params:jmap:submission"],
         |  "methodCalls": [[
         |    "Identity/get",
         |    {
         |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
         |      "ids": ["$id"]
         |    },
         |    "c1"]]
         |}""".stripMargin

    val response =  `given`
      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
      .body(request)
    .when
      .post.prettyPeek()
    .`then`
      .statusCode(SC_OK)
      .contentType(JSON)
      .extract
      .body
      .asString
```
Response:
```
{
    "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
    "methodResponses": [
        [
            "error",
            {
                "type": "serverFail",
                "description": "JsError.get"
            },
            "c1"
        ]
    ]
}
```
I will dig into this tomorrow afternoon.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 5]
```diff
@@ -663,4 +670,304 @@ trait IdentitySetContract {
...
+
+  @Test
+  def destroyShouldHandleMixedCases(server: GuiceJamesServer): Unit = {
+    val customId1 = SMono(server.getProbe(classOf[IdentityProbe])
+      .save(BOB, IdentityCreationRequest(name = Some(IdentityName("Bob (custom address 1)")),
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:894]: Should we create a sample val and re-use it,  less boilerplate
`val SAMPLE_CREATTION_REQUEST = IdentityCreationRequest(....`

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 6]
```diff
@@ -663,4 +670,304 @@ trait IdentitySetContract {
...
+         |				"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |				"destroy": ["$customId1", "$customId2", "$defaultServerSetIdentity"]
+         |			},
+         |			"c1"
+         |		]
+         |	]
+         |}""".stripMargin
+
+    val response: String = `given`
+      .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:929]: Declare it in `BeforeEach` def, then we don't need to redeclare it for each time

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala (hunk 5 above)
Comment (quantranhong1999) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:894]: Ok, I think I can declare it as an object's constant.

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala (hunk 6 above)
Comment (quantranhong1999) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:929]: Why do you comment on this line? What declaration do you mean?

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:929]: I think he is suggesting adding the header in the setup method, in the RestAussured `requestSpecification` init

**Comment by Arsnael** [comment]
You can rebase on master btw :)

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala (hunk 6 above)
Comment (vttranlina) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:929]: My mean
```scala
    requestSpecification = baseRequestSpecBuilder(server)
      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
      .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
      .build
```
then, we don't need to add a rfc8621 header  in each `given`

**Comment by quantranhong1999** [comment]
Rebased with https://github.com/apache/james-project/pull/756/commits/910c3c57e64e15f5295c519355aae0f355b1064d and https://github.com/apache/james-project/pull/756/commits/bf2f2d8a15f185a53e93fc05a6fcfe2a099c32d7 solve Rene and Tung comments

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]
Can you squash @quantranhong1999 ?

**Pull request merged**